### PR TITLE
Fix #108: Represent stack locations

### LINKED ISSUE #108: Represent stack locations
Real stacks have module/file/function information. Our "stacks" have only file/line/column information (and a message which could repurposed to contain the missing information).

Example:
`xunit.execution.desktop.dll!Xunit.Sdk.TestClassRunner<Xunit.Sdk.IXunitTestCase>.RunTestMethodsAsync() Line 205  C#`

Need:
- module
- fully qualified method name (or other scope, e.g. field initializer)
- file location/code offset from start of function
- message (already have)
- a way to express "external code"/"unresolved frame"

### PR DESCRIPTION
**Pull request by ghost**
@michaelcfanning

### CODE REVIEW

File: document/CoreFormat.mdk [hunk 1]
```diff
@@ -1003,30 +1003,16 @@ relevant to a single result.
...
+A `result` object may contain a property named `stacks` whose value is an array of one or more
+unique (&sect;[#unique-array-properties]) `stack` objects (&sect;[#stack]).
 The `stacks` property is intended for use by analysis tools that collects call stack information 
 in the process of producing results.
 
-Each `stack` object in the stacks array shall describe a single call stack.
-A call stack is a sequence of nested function calls.
-A stack is in turn represented by an array of `annotatedCodeLocation` objects.
-This array shall include every function call in the stack for which the tool has information,
-and the entries that are present shall occur in order with the innermost call first and the outermost call last.
```
Comment (ghost) [inline document/CoreFormat.mdk:1017]: This paragraph moved down (in slightly modified form) into the `stack` section.

File: document/CoreFormat.mdk [hunk 2]
```diff
@@ -1743,13 +1729,138 @@ A `codeFlow` object may contain a property named `tags`, whose value is a tag se
...
+### `message` property { #stack-message }
+
+A `stack` object may contain a property named `message` whose value is a string
+containing a message relevant to this call stack.
+
+### `frames` property { #stack-frames }
+
+A `stack` object shall contain a property named `frames` whose value is an array of one or more `stackFrame` objects (&sect;[#stackFrame]).
+This array shall include every function call in the stack for which the tool has information,
+and the entries that are present shall occur in order with the innermost call first and the outermost call last.
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: do you think you should make the chronological relationship of these things clear or will everyone grok it?

File: document/CoreFormat.mdk [hunk 3]
```diff
@@ -1743,13 +1729,138 @@ A `codeFlow` object may contain a property named `tags`, whose value is a tag se
...
+The entries in this array need not be unique.
+It is possible for the same frame to occur multiple times if the calls stack includes a recursion.
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: not 'calls stack', s/be 'call stack'

File: document/CoreFormat.mdk [hunk 4]
```diff
@@ -1743,13 +1729,138 @@ A `codeFlow` object may contain a property named `tags`, whose value is a tag se
...
+
+A `stackFrame` object may contain a property named `message` whose value is a string
+containing a message relevant to this stack frame.
+
+### `uri` property { #stackFrame-uri }
+
+A `stackFrame` object may contain a property named `uri` whose value is a string
+containing the URI of the source code file to which this stack frame refers.
+
+~ IsoNote
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: do we really need to make this explicit? aren't we just saying this information is absent if we don't have it?

File: document/CoreFormat.mdk [hunk 5]
```diff
@@ -1743,13 +1729,138 @@ A `codeFlow` object may contain a property named `tags`, whose value is a tag se
...
+containing the name of the module that contains the location to which this stack frame refers.
+
+### `fullyQualifiedLogicalName` property { #stackFrame-fullyQualifiedLogicalName }
+
+A `stackFrame` object may contain a property named `fullyQualifiedLogicalName` whose value
+is a string containing the fully qualified name of the method to which this stack frame refers.
+See &sect;[#location-fullyQualifiedLogicalName] for examples.
+
+~ IsoNote
+The `fullyQualifiedLogicalName` property will be absent in the analysis tool does not
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: there is a typo in this statement s/be 'if the analysis tool' not 'in the analysis tool'. the statement is also not correct. a managed stack frame has no symbol information but constructs a fully qualified name (due to the discoverability of IL).

File: document/CoreFormat.mdk [hunk 6]
```diff
@@ -1743,13 +1729,138 @@ A `codeFlow` object may contain a property named `tags`, whose value is a tag se
...
+
+A `stackFrame` object may contain a property named `logicalLocationKey` whose value is a string.
+If present, this string shall be equal to the name of one of the properties on the `run.logicalLocations` object (&sect;[#run-logicalLocations]),
+which provides additional information about the logical location specified by `fullyQualifiedLogicalName` (&sect;[#stackFrame-fullyQualifiedLogicalName]).
+For more information about the purpose of this property, see &sect;[#location-fullyQualifiedLogicalName].
+
+### `address` { #stackFrame-address }
+
+A `stackFrame` object may contain a property named `address` whose value is an integer
+containing the byte offset of the location represented by this stack frame from the start
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: I'm not sure this is right. Typically a stack frame address is the actual address of the function (+ offset ), not rendered as the offset from base address of the module. This reflects the literal execution of the code (where the return address for execution is pushed onto the stack

File: document/CoreFormat.mdk [hunk 7]
```diff
@@ -1743,13 +1729,138 @@ A `codeFlow` object may contain a property named `tags`, whose value is a tag se
...
+of the module containing this location.
+
+### `offset` { #stackFrame-offset }
+
+A `stackFrame` object may contain a property named `offset` whose value is an integer
+containing the byte offset of the location represented by this stack frame from the start
+of the method represented by this stack frame.
+
+### `parameters` { #stackFrame-parameters }
+
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk:1827]: would prefer to have example use on this from actual log files, btw, before we finalize it (or add to v1). all other members of this data type are exercised by current converters/direct producers.

File: document/CoreFormat.mdk (hunk 4 above)
Comment (ghost) [inline document/CoreFormat.mdk]: It's saying a little more than that: it's explaining _why_ you might not have it. I'd like to keep this (and the similar not on `line` just below).

File: document/CoreFormat.mdk (hunk 3 above)
Comment (ghost) [inline document/CoreFormat.mdk]: fixed

File: document/CoreFormat.mdk (hunk 2 above)
Comment (ghost) [inline document/CoreFormat.mdk]: I think "in order with the innermost call first and the outermost call last" is sufficient.

File: document/CoreFormat.mdk (hunk 4 above)
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: I find it odd that you don't mention the obvious clarifying case, though, the presence or non-presence of pdbs/symbols. note that all of this case is not the primary use case. many static analysis tools will produce speculative stacks working strictly against source code. the address/offset may be immaterial for these cases..

File: document/CoreFormat.mdk (hunk 5 above)
Comment (ghost) [inline document/CoreFormat.mdk]: ok, I'll omit this note.

File: document/CoreFormat.mdk (hunk 6 above)
Comment (ghost) [inline document/CoreFormat.mdk]: oh, right. How about:

> A `stackFrame` object may contain a property named `address` whose value is an integer
> containing the address in memory of the location represented by this stack frame.

File: document/CoreFormat.mdk (hunk 2 above)
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: disagree. 'most recent' is entirely clear. 'innermost' requires some mutually agreed on mental model that might not coincide with the readers.

File: document/CoreFormat.mdk (hunk 7 above)
Comment (ghost) [inline document/CoreFormat.mdk:1827]: I assume your comment refers to `stackFrame.parameters` in particular and not to `stackFrame` as a whole. What do you want to do? Do we have a tool that has access to this information? Do you want to omit it until we do? As the schema stands today, this member is present, and I was just documenting what was in the schema.

File: document/CoreFormat.mdk (hunk 6 above)
Comment (michaelcfanning) [inline document/CoreFormat.mdk]: perfect

File: document/CoreFormat.mdk (hunk 7 above)
Comment (michaelcfanning) [inline document/CoreFormat.mdk:1827]: yes, only 'parameters'. I will send mail trying to solicit relevant log files, will include you. we can leave things as is for now.

File: document/CoreFormat.mdk (hunk 2 above)
Comment (ghost) [inline document/CoreFormat.mdk]: ok. "... in chronological order with the most recent (innermost) call first and the least recent (outermost) call last."

File: document/CoreFormat.mdk (hunk 4 above)
Comment (ghost) [inline document/CoreFormat.mdk]: ok, I think the simplest thing is to go by your initial comment and just omit the notes on `uri` and `line`.

**Pull request merged**